### ISSUE TITLE
Request: Ability to add items to the queue

### ISSUE THREAD

**Issue by maddox** [jishi/node-sonos-http-api#357]
It'd be super awesome to be able to queue items. 

My reasoning is that I'd like to take a snapshot of the queue so that I can restore it later. We already have `/queue` to get the snapshot. But there's no way to feed it back and and re-queue everything.

It'd be awesome if we could do that.

**Comment by jishi** [comment]
Comparing this to the possibility to create a playlist of the queue and restore that playlist, which one is preferred? Is there any advantage to have an external snapshot compared to containing it within sonos functionality?

The queue as is doesn't contain enough info to restore it, but that is fairly trivial to add. Restoring it would best then require a POST of the exact json body in return, which probably is easy to integrate but harder to test. How would you invoke that? Just as reference.

**Comment by maddox** [comment]
To be honest, both would be nice, lol. But what I really am after is a way to stash the current state away, create a new queue (with a favorite or playlist), play some, then completely restore state back.

Being able to completely restore state would include everything. Volume, currently playing item and it's play state (progress through song). And being able to do this on individual zones would be best.

I'm trying to create a good party mode function so I instantly set a new playlist, play it, join zones, etc. And then be able to instantly roll back.

Maybe however the re-queuing works, could be abstracted and exposed in a way that we could manually add items to a queue as well.

**Comment by jishi** [comment]
Comparing this to the possibility to create a playlist of the queue and
restore that playlist, which one is preferred? Is there any advantage to
have an external snapshot compared to containing it within sonos
functionality?

The queue as is doesn't contain enough info to restore it, but that is
fairly trivial to add. Restoring it would best then require a POST of the
exact json body in return, which probably is easy to integrate but harder
to test. How would you invoke that? Just as reference.

**Comment by maddox** [comment]
I think your message got screwed up there?

**Comment by jishi** [comment]
There was some issue with the email replies last week, it was a 4 days old email that finally arrived :)

**Comment by maddox** [comment]
But that reply is up👆there. I replied to it 😝

**Comment by jishi** [comment]
Yeah, because I noticed that the email didn't go through, I answered directly at GitHub with the same content as the email, hence the confusion...

**Comment by maddox** [comment]
Please scroll up. I have replied to your message to further this conversation. Your message actually went through 4 days ago. I got it and replied to it.

**Comment by jishi** [comment]
Yeah, no, I know. I was just explaining why it got repeated. There is no real confusion :)

**Comment by maddox** [comment]
So what do you think about what I said here: https://github.com/jishi/node-sonos-http-api/issues/357#issuecomment-255817047

**Comment by jishi** [comment]
Ah sorry, didn't understand that you wanted an answer to your comment. 

Restoring state with presets actually work today, excluding restoring a queue. I use that for the say/clip commands, where I stash away a preset of the systems current state, then restoring that. In that case, the queue is left intact (since I don't replace it), so it is possible to just switch back to queue and restore trackNo and time of track. Restoring a queue will take some time if it is big (+100 tracks) though, so this is probably not something you want to do on a regular basis.

I think the properties for restoring queue-position and trackNo/time is undocumented, but if the preset contains:

`trackNo`
`elapsedTime` (in seconds)

It would use them. Also, queue playback is triggered by having the avtransporturi set to `x-rincon-queue:[uuid of player]` which can be set by the `uri` property. It is however not possible to get the current AVTransportURI easily via the api. 

Just allowing dumping the queue with metadata (required for restoring) and having an action where you could POST this back would be fairly trivial to fix. This in combination with dynamic presets would solve your need. I'll look into it when I find the time for it!

**Comment by jishi** [comment]
Hm, I thought I had responded to this. Seems like the required metadata for each track in the queue isn't readily available, which makes this much harder than I anticipated. Dumping the queue to a temporary playlist would probably be easier.

There are some methods for saving "state", but I haven't understood how they work yet, doesn't seem to be used by the official controller very often, might be related to saving state during an upgrade or something.

**Comment by maddox** [comment]
> Hm, I thought I had responded to this. Seems like the required metadata for each track in the queue isn't readily available, which makes this much harder than I anticipated

yeah, the python SoCo library has this same issue. I figured it'd be the same for this one. So I figured just serializing the queue down and reloading it might be easier.
